### PR TITLE
New Module win_dns_delegation

### PR DESCRIPTION
**Pull request by sgruber94**
### SUMMARY
The win_dns_delegation module manage a Windows Server DNS Zone Delegations

### ISSUE TYPE
New Module Pull Request

### COMPONENT NAME
win_dns_delegation

### ADDITIONAL INFORMATION
Includes Integration Tests - See documentation file for full details on how the module works.

### CODE REVIEW

**Comment by sgruber94** [comment]
Tests failed, because example.com zone is not  in test env. Is there any favorite DNS zone, I should use?

File: plugins/modules/win_dns_delegation.ps1 [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+# SPDX-License-Identifier: GPL-3.0-only
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$spec = @{
+    options             = @{
+        name       = @{ type = "str"; required = $true }
+        zone       = @{ type = "str"; required = $true }
+        state      = @{ type = "str"; choices = "absent", "present"; default = "present" }
+        NameServer = @{ type = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:12]: Keep options in `snake_case` format, like `name_server`.

File: plugins/modules/win_dns_delegation.ps1 [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        IPAddress  = @{ type = "list"; elements = "str" }
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:13]: Ditto, use `snake_case` for module options; `ip_address`.

File: plugins/modules/win_dns_delegation.ps1 [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    Import-Module DNSServer
+} Catch {
+    # Couldn't load the DNS Module
+    $module.FailJson("The DNSServer module failed to load properly: $($_.Exception.Message)", $_)
+}
+
+try {
+    $delegationzone = Get-DnsServerZoneDelegation @parms
+} catch {
+    $module.FailJson("Failed to get dns delegation zone. $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:40]: There's not much need to wrap this in a try catch, Ansible will automatically pick up uncaught exception.

File: plugins/modules/win_dns_delegation.ps1 [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    $parms.IPAddress = $IPAddress
+    if ($check_mode) { $parms.WhatIf = $check_mode }
+    #check if exist
+    if ($delegationzone) {
+        if ($delegationzone[0].NameServer.RecordData.NameServer.trim(".") -notmatch $nameserver) {
+            try {
+                Add-DnsServerZoneDelegation @parms -PassThru -Confirm:$false
+                $module.Result.changed = $true
+            } catch {
+                $module.FailJson("Failed to add dns delegation zone $($name): $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:54]: Same here, the try catch isn't really doing much here, just best off leaving it out.

File: plugins/modules/win_dns_delegation.ps1 [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        } else {
+            #entry exist need to compare
+            $addressv4 = $delegationzone[0].IPAddress.RecordData.IPV4Address.IPAddressToString
+            $addressv6 = $delegationzone[0].IPAddress.RecordData.IPV6Address.IPAddressToString
+            if ($addressv4) {
+                try {
+                    $diffnsipv4 = Compare-Object -ReferenceObject $addressv4 -DifferenceObject $ipaddress
+                } catch {
+                    $module.FailJson("Failed to compare delegation NameServer zone $($name): $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:65]: Ditto, remove try/catch

File: plugins/modules/win_dns_delegation.ps1 [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                }
+            }
+            if ($addressv6) {
+                try {
+                    $diffnsipv6 = Compare-Object -ReferenceObject $addressv6 -DifferenceObject $ipaddress
+                } catch {
+                    $module.FailJson("Failed to compare delegation NameServer zone $($name): $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:72]: Ditto, remove try/catch

File: plugins/modules/win_dns_delegation.ps1 [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                }
+            }
+            if (($diffnsipv4.count -gt 0) -or ($diffnsipv6.count -gt 0 )) {
+                try {
+                    #changing ip address
+                    Set-DnsServerZoneDelegation @parms -Confirm:$false
+                    $module.Result.changed = $true
+                } catch {
+                    $module.FailJson("Failed to change dns Server zone $($name): $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:81]: Ditto, remove try/catch

File: plugins/modules/win_dns_delegation.ps1 [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                }
+            } else {
+                $module.Result.changed = $false
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:84]: Just an FYI, `changed = $false` is the default, you typically only ever need to set this value if you are setting it to `$true`.

File: plugins/modules/win_dns_delegation.ps1 [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            }
+        }
+    } else {
+        #create dns delegation
+        try {
+            Add-DnsServerZoneDelegation @parms -PassThru -Confirm:$false
+            $module.Result.changed = $true
+        } catch {
+            $module.FailJson("Failed to add dns delegation zone $($name): $($_.Exception.Message)", $_)
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:93]: Ditto, remove try/catch

File: plugins/modules/win_dns_delegation.ps1 [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+try {
+    $delegationzone = Get-DnsServerZoneDelegation @parms
+} catch {
+    $module.FailJson("Failed to get dns delegation zone. $($_.Exception.Message)", $_)
+}
+
+if ($state -eq "present") {
+    $parms.NameServer = $NameServer
+    $parms.IPAddress = $IPAddress
+    if ($check_mode) { $parms.WhatIf = $check_mode }
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.ps1:46]: You may as well add this to `$parms` when you define it as it's used for both state present and absent. A switch can be set to `$false` as well to indicate it's not set allowing you to do it in all cases and not just in check mode.

File: plugins/modules/win_dns_delegation.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021 Sebastian Gruber(@sgruber94) ,dacoso GmbH All Rights Reserved.
+# SPDX-License-Identifier: GPL-3.0-only
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+DOCUMENTATION = r'''
+---
+module: win_dns_delegation
```
Comment (jborean93) [inline plugins/modules/win_dns_delegation.py:9]: Needs `version_added: '1.4.0'.

File: tests/integration/targets/win_dns_delegation/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    include_management_tools: true
+    include_sub_features: true
+    state: present
+  register: ensure_addns_roles
+
+- name: Reboot
+  ansible.windows.win_reboot:
+  when: ensure_addns_roles.reboot_required
+
+- name: ADD | DNS Zone Delegation
```
Comment (jborean93) [inline tests/integration/targets/win_dns_delegation/tasks/main.yml:15]: This is a very bare set of tests, you should be testing all sorts of scenarios like check mode, state present/absent, idempotency. Not just the modules you should also be checking the results to make sure the module did what you told it to do.

**Review by jborean93 — CHANGES_REQUESTED** [review]

**Comment by jborean93** [comment]
For the test failure this is all standalone, you need to setup what is needed to run them.

**Comment by jborean93** [comment]
See https://github.com/ansible-collections/community.windows/pull/194#issuecomment-877442730

**Pull request closed without merging**